src/features/edge-intersection/intersection-finder.ts

```diff
@@ -0,0 +1,98 @@
+/********************************************************************************
+ * Copyright (c) 2021 EclipseSource and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable } from "inversify";
+import TinyQueue from "tinyqueue";
+import { Point } from "../../utils/geometry";
+import { EdgeRouting, IEdgeRoutePostprocessor, RoutedPoint } from "../routing/routing";
+import { addRoute, checkWhichEventIsLeft, runSweep, SweepEvent } from "./sweepline";
+
+export interface Intersection {
+    readonly routable1: string;
+    readonly segmentIndex1: number;
+    readonly routable2: string;
+    readonly segmentIndex2: number;
+    readonly intersectionPoint: Point;
+}
+
+export interface IntersectingRoutedPoint extends RoutedPoint {
+    intersections: Intersection[];
+}
+
+export function isIntersectingRoutedPoint(routedPoint: Point): routedPoint is IntersectingRoutedPoint {
+    return routedPoint !== undefined && 'intersections' in routedPoint && 'kind' in routedPoint;
+}
+
+export const BY_X_THEN_Y = (a: Intersection, b: Intersection): number => {
+    if (a.intersectionPoint.x === b.intersectionPoint.x) {
+        return a.intersectionPoint.y - b.intersectionPoint.y;
+    }
+    return a.intersectionPoint.x - b.intersectionPoint.x;
+};
+
+
+export class RouteSegment {
+    constructor(public routableId: string, public segmentIndex: number, public point: Point) { }
+}
```

Comment:
General note: I'd propose to use plain objects for such data structures, described by an `interface` type, instead of a `class`. This applies to other data structures in Sprotty as well, e.g. actions. Maybe we should refactor those at some point.

Rationale: The concept of using classes for data structures comes from the influence the many years of Java development have left on us (at least I can tell that for me and @JanKoehnlein). But this is rather unnatural in the JavaScript world, where plain objects are predominant. I would still use classes for service implementations and structures with a lot of attached behavior like the internal SModel, though.

Reply:
Yeah, you're totally right. I tend to fall back to classes inadvertently.
Actually, however, I realized that `RouteSegment` is actually never used, so this is a left-over. Sorry about that.

Reply:
:+1: I have created a separate issue (https://github.com/eclipse/sprotty/issues/233) to track the potential refactoring of action classes to `interfaces`: